### PR TITLE
SAA-433: Cater for cancellations when creating attendance records

### PR DESCRIPTION
**Pull request by leejacobson-moj**
## Overview

Adds support for creating attendance records for scheduled instances which have been cancelled

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesService.kt [hunk 1]
```diff
@@ -73,10 +73,23 @@ class ManageAttendancesService(
...
+            attendanceReason = attendanceReasonRepository.findByCode(AttendanceReasonEnum.NOT_REQUIRED),
+            recordedTime = LocalDateTime.now(),
+            recordedBy = ServiceName.SERVICE_NAME.value,
+          )
         } ?: Attendance(
         scheduledInstance = instance,
         prisonerNumber = allocation.prisonerNumber,
         payAmount = instance.rateFor(allocation.payBand),
+        status = if (instance.cancelled) AttendanceStatus.COMPLETED else AttendanceStatus.WAITING,
       )
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesService.kt:93]: Is the `instance.takeIf { it.cancelled }` already filtering just to cancelled? So is the `status =  if ( cancelled) ` necessary?

Comment (leejacobson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesService.kt:93]: Yes, you're right – thanks for spotting this Tim. I was suppose to remove this line.

**Review by timharrison-moj — DISMISSED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesService.kt [hunk 2]
```diff
@@ -73,6 +73,18 @@ class ManageAttendancesService(
             recordedTime = LocalDateTime.now(),
             recordedBy = ServiceName.SERVICE_NAME.value,
           )
+        } ?: instance.takeIf { it.cancelled }
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesService.kt:76]: I am wondering if this block of code (now it has grown) might be easier to read in a when statement? For example

```
      val attendance = when {
        allocation.status(PrisonerStatus.AUTO_SUSPENDED, PrisonerStatus.SUSPENDED) -> Attendance(...)
        instance.cancelled -> Attendance(...)
        else -> Attendance(...)
      }

```

Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesService.kt:76]: I agree with this - despite just approving - it did take me a while to look through that particular block of code.

**Review by timharrison-moj — APPROVED** [review]
LGTM

**Pull request merged**